From: SDK conformance patches for MSVC v145 (toolset bump in a4c3a58)
Subject: Patch legacy Unreal/Deus Ex SDK headers for strict modern C++ conformance

These files live under gitignored games/ and ship with the vendor SDK drop.
Apply from repo root with:  patch -p1 < patches/sdk-v145-compat.patch

UnTemplate.h: add 'typename' to 7 dependent-type ConstInitType uses.
Window.h:     add '&Class::' to 23 member-function-pointer casts and
              const-qualify one ParseToken cast.

--- a/games/DeusEx/core/inc/UnTemplate.h
+++ b/games/DeusEx/core/inc/UnTemplate.h
@@ -1343,7 +1343,7 @@
 		INT HashNext;
 		TK Key;
 		TI Value;
-		TPair( TTypeInfo<TK>::ConstInitType InKey, TTypeInfo<TI>::ConstInitType InValue )
+		TPair( typename TTypeInfo<TK>::ConstInitType InKey, typename TTypeInfo<TI>::ConstInitType InValue )
 		: Key( InKey ), Value( InValue )
 		{}
 		TPair()
@@ -1385,7 +1385,7 @@
 		Rehash();
 		unguardSlow;
 	}
-	TI& Add( TTypeInfo<TK>::ConstInitType InKey, TTypeInfo<TI>::ConstInitType InValue )
+	TI& Add( typename TTypeInfo<TK>::ConstInitType InKey, typename TTypeInfo<TI>::ConstInitType InValue )
 	{
 		guardSlow(TMapBase::Add);
 		TPair& Pair   = *new(Pairs)TPair( InKey, InValue );
@@ -1448,7 +1448,7 @@
 		Rehash();
 		unguardSlow;
 	}
-	TI& Set( TTypeInfo<TK>::ConstInitType InKey, TTypeInfo<TI>::ConstInitType InValue )
+	TI& Set( typename TTypeInfo<TK>::ConstInitType InKey, typename TTypeInfo<TI>::ConstInitType InValue )
 	{
 		guardSlow(TMap::Set);
 		for( INT i=Hash[(GetTypeHash(InKey) & (HashCount-1))]; i!=INDEX_NONE; i=Pairs(i).HashNext )
@@ -1457,7 +1457,7 @@
 		return Add( InKey, InValue );
 		unguardSlow;
 	}
-	INT Remove( TTypeInfo<TK>::ConstInitType InKey )
+	INT Remove( typename TTypeInfo<TK>::ConstInitType InKey )
 	{
 		guardSlow(TMapBase::Remove);
 		INT Count=0;
@@ -1558,18 +1558,18 @@
 				new(Values)TI(Pairs(i).Value);
 		unguardSlow;
 	}
-	TI& Add( TTypeInfo<TK>::ConstInitType InKey, TTypeInfo<TI>::ConstInitType InValue )
+	TI& Add( typename TTypeInfo<TK>::ConstInitType InKey, typename TTypeInfo<TI>::ConstInitType InValue )
 	{
 		return TMapBase<TK,TI>::Add( InKey, InValue );
 	}
-	TI& AddUnique( TTypeInfo<TK>::ConstInitType InKey, TTypeInfo<TI>::ConstInitType InValue )
+	TI& AddUnique( typename TTypeInfo<TK>::ConstInitType InKey, typename TTypeInfo<TI>::ConstInitType InValue )
 	{
 		for( INT i=Hash[(GetTypeHash(InKey) & (HashCount-1))]; i!=INDEX_NONE; i=Pairs(i).HashNext )
 			if( Pairs(i).Key==InKey && Pairs(i).Value==InValue )
 				return Pairs(i).Value;
 		return Add( InKey, InValue );
 	}
-	INT RemovePair( TTypeInfo<TK>::ConstInitType InKey, TTypeInfo<TI>::ConstInitType InValue )
+	INT RemovePair( typename TTypeInfo<TK>::ConstInitType InKey, typename TTypeInfo<TI>::ConstInitType InValue )
 	{
 		guardSlow(TMap::Remove);
 		INT Count=0;
--- a/games/DeusEx/window/inc/Window.h
+++ b/games/DeusEx/window/inc/Window.h
@@ -1585,7 +1585,7 @@
 	WUrlButton()
 	{}
 	WUrlButton( WWindow* InOwner, const TCHAR* InURL, INT InId=0 )
-	: WCoolButton( InOwner, InId, FDelegate(this,(TDelegate)OnClick) )
+	: WCoolButton( InOwner, InId, FDelegate(this,(TDelegate)&WUrlButton::OnClick) )
 	, URL( InURL )
 	{
 		FrameFlags = CBFF_ShowOver | CBFF_UrlStyle | CBFF_NoCenter;
@@ -2278,7 +2278,7 @@
 			Exec->Exec( TEXT("TakeFocus"), *GLogWindow );
 			TCHAR NewURL[1024];
 			if
-			(	ParseToken(*(TCHAR**)&CD->lpData,NewURL,ARRAY_COUNT(NewURL),0)
+			(	ParseToken(*(const TCHAR**)&CD->lpData,NewURL,ARRAY_COUNT(NewURL),0)
 			&&	NewURL[0]!='-')
 				Exec->Exec( *(US+TEXT("Open ")+NewURL),*GLogWindow );
 		}
@@ -2485,8 +2485,8 @@
 	// Constructor.
 	WPasswordDialog()
 	: WDialog	  ( TEXT("PasswordDialog"), IDDIALOG_Password )
-	, OkButton    ( this, IDOK,     FDelegate(this,(TDelegate)EndDialogTrue) )
-	, CancelButton( this, IDCANCEL, FDelegate(this,(TDelegate)EndDialogFalse) )
+	, OkButton    ( this, IDOK,     FDelegate(this,(TDelegate)&WDialog::EndDialogTrue) )
+	, CancelButton( this, IDCANCEL, FDelegate(this,(TDelegate)&WDialog::EndDialogFalse) )
 	, Name		  ( this, IDEDIT_Name )
 	, Password	  ( this, IDEDIT_Password )
 	, Prompt      ( this, IDLABEL_Prompt )
@@ -2535,7 +2535,7 @@
 	WTextScrollerDialog( const TCHAR* InCaption, const TCHAR* InMessage )
 	: WDialog	( TEXT("TextScrollerDialog"), IDDIALOG_TextScroller )
 	, TextEdit  ( this, IDEDIT_TextEdit )
-	, OkButton  ( this, IDOK, FDelegate(this,(TDelegate)EndDialogTrue) )
+	, OkButton  ( this, IDOK, FDelegate(this,(TDelegate)&WDialog::EndDialogTrue) )
 	, Caption   ( InCaption )
 	, Message   ( InMessage )
 	{}
@@ -3648,8 +3648,8 @@
 				Rect.Min.X = 28+OwnerProperties->GetDividerWidth();
 				Rect.Min.Y--;
 				TrackControl = new WTrackBar( &OwnerProperties->List );
-				TrackControl->ThumbTrackDelegate    = FDelegate(this,(TDelegate)OnTrackBarThumbTrack);
-				TrackControl->ThumbPositionDelegate = FDelegate(this,(TDelegate)OnTrackBarThumbPosition);
+				TrackControl->ThumbTrackDelegate    = FDelegate(this,(TDelegate)&FPropertyItem::OnTrackBarThumbTrack);
+				TrackControl->ThumbPositionDelegate = FDelegate(this,(TDelegate)&FPropertyItem::OnTrackBarThumbPosition);
 				TrackControl->OpenWindow( 0 );
 				TrackControl->SetTicFreq( 32 );
 				TrackControl->SetRange( 0, 255 );
@@ -3675,8 +3675,8 @@
 
 				ComboControl = new WComboBox( HolderControl );
 				ComboControl->Snoop = this;
-				ComboControl->SelectionEndOkDelegate     = FDelegate(this,(TDelegate)ComboSelectionEndOk);
-				ComboControl->SelectionEndCancelDelegate = FDelegate(this,(TDelegate)ComboSelectionEndCancel);
+				ComboControl->SelectionEndOkDelegate     = FDelegate(this,(TDelegate)&FPropertyItem::ComboSelectionEndOk);
+				ComboControl->SelectionEndCancelDelegate = FDelegate(this,(TDelegate)&FPropertyItem::ComboSelectionEndCancel);
 				ComboControl->OpenWindow( 0 );
 				ComboControl->MoveWindow( Rect-HolderRect.Min, 1 );
 
@@ -3718,29 +3718,29 @@
 			{
 				if( Expandable )
 				{
-					AddButton( LocalizeGeneral("AddButton",TEXT("Window")), FDelegate(this,(TDelegate)OnArrayAdd) );
+					AddButton( LocalizeGeneral("AddButton",TEXT("Window")), FDelegate(this,(TDelegate)&FPropertyItem::OnArrayAdd) );
 				}
-				AddButton( LocalizeGeneral("EmptyButton",TEXT("Window")), FDelegate(this,(TDelegate)OnArrayEmpty) );
+				AddButton( LocalizeGeneral("EmptyButton",TEXT("Window")), FDelegate(this,(TDelegate)&FPropertyItem::OnArrayEmpty) );
 			}
 			if( Cast<UArrayProperty>(Property->GetOuter()) )
 			{
 				if( Parent->Expandable )
 				{
-					AddButton( LocalizeGeneral("InsertButton",TEXT("Window")), FDelegate(this,(TDelegate)OnArrayInsert) );
-					AddButton( LocalizeGeneral("DeleteButton",TEXT("Window")), FDelegate(this,(TDelegate)OnArrayDelete) );
+					AddButton( LocalizeGeneral("InsertButton",TEXT("Window")), FDelegate(this,(TDelegate)&FPropertyItem::OnArrayInsert) );
+					AddButton( LocalizeGeneral("DeleteButton",TEXT("Window")), FDelegate(this,(TDelegate)&FPropertyItem::OnArrayDelete) );
 				}
 			}
 			if( Property->IsA(UStructProperty::StaticClass()) && appStricmp(Cast<UStructProperty>(Property)->Struct->GetName(),TEXT("Color"))==0 )
 			{
 				// Color.
-				AddButton( LocalizeGeneral("BrowseButton",TEXT("Window")), FDelegate(this,(TDelegate)OnChooseColorButton) );
+				AddButton( LocalizeGeneral("BrowseButton",TEXT("Window")), FDelegate(this,(TDelegate)&FPropertyItem::OnChooseColorButton) );
 			}
 			else if( Property->IsA(UObjectProperty::StaticClass()) )
 			{
 				// Class.
-				AddButton( LocalizeGeneral("BrowseButton",TEXT("Window")), FDelegate(this,(TDelegate)OnBrowseButton) );
-				AddButton( LocalizeGeneral("UseButton",   TEXT("Window")), FDelegate(this,(TDelegate)OnUseCurrentButton) );
-				AddButton( LocalizeGeneral("ClearButton", TEXT("Window")), FDelegate(this,(TDelegate)OnClearButton) );
+				AddButton( LocalizeGeneral("BrowseButton",TEXT("Window")), FDelegate(this,(TDelegate)&FPropertyItem::OnBrowseButton) );
+				AddButton( LocalizeGeneral("UseButton",   TEXT("Window")), FDelegate(this,(TDelegate)&FPropertyItem::OnUseCurrentButton) );
+				AddButton( LocalizeGeneral("ClearButton", TEXT("Window")), FDelegate(this,(TDelegate)&FPropertyItem::OnClearButton) );
 			}
 			if
 			(	(Property->IsA(UFloatProperty ::StaticClass()))
@@ -4196,8 +4196,8 @@
 		if( PersistentName!=NAME_None )
 			GConfig->GetInt( TEXT("WindowPositions"), *(FString(*PersistentName)+TEXT(".Split")), DividerWidth );
 		PropertiesWindows.AddItem( this );
-		List.DoubleClickDelegate     = FDelegate(this,(TDelegate)OnListDoubleClick);
-		List.SelectionChangeDelegate = FDelegate(this,(TDelegate)OnListSelectionChange);
+		List.DoubleClickDelegate     = FDelegate(this,(TDelegate)&WProperties::OnListDoubleClick);
+		List.SelectionChangeDelegate = FDelegate(this,(TDelegate)&WProperties::OnListSelectionChange);
 		unguard;
 	}
 
@@ -4891,7 +4891,7 @@
 	void OnItemSetFocus()
 	{
 		FPropertyItemBase::OnItemSetFocus();
-		AddButton( LocalizeGeneral("DefaultsButton",TEXT("Window")), FDelegate(this,(TDelegate)OnResetToDefaultsButton) );
+		AddButton( LocalizeGeneral("DefaultsButton",TEXT("Window")), FDelegate(this,(TDelegate)&FObjectConfigItem::OnResetToDefaultsButton) );
 	}
 	void Expand()
 	{
@@ -5114,10 +5114,10 @@
 	WWizardDialog()
 	: WDialog	    ( TEXT("WizardDialog"), IDDIALOG_WizardDialog )
 	, PageHolder    ( this, IDC_PageHolder )
-	, BackButton    ( this, IDC_Back,   FDelegate(this,(TDelegate)OnBack  ) )
-	, NextButton    ( this, IDC_Next,   FDelegate(this,(TDelegate)OnNext  ) )
-	, FinishButton  ( this, IDC_Finish, FDelegate(this,(TDelegate)OnFinish) )
-	, CancelButton  ( this, IDCANCEL,   FDelegate(this,(TDelegate)OnCancel) )
+	, BackButton    ( this, IDC_Back,   FDelegate(this,(TDelegate)&WWizardDialog::OnBack  ) )
+	, NextButton    ( this, IDC_Next,   FDelegate(this,(TDelegate)&WWizardDialog::OnNext  ) )
+	, FinishButton  ( this, IDC_Finish, FDelegate(this,(TDelegate)&WWizardDialog::OnFinish) )
+	, CancelButton  ( this, IDCANCEL,   FDelegate(this,(TDelegate)&WWizardDialog::OnCancel) )
 	, Pages         ()
 	, CurrentPage   ( NULL )
 	{}
